### PR TITLE
BUG: Fix size-checking in masked_where, and structured shrink_mask

### LINKED ISSUE #4520: Strange behavior of np.ma.masked_where
When the mask and array shapes do not match and the mask is always false, no exception is thrown, probably a warning could be good. Consider the example

```
x = np.arange(10)
y = np.arange(5) 
z = np.ma.masked_where(y > 6, x)
```

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #4520

### CODE REVIEW

File: numpy/ma/core.py [hunk 1]
```diff
@@ -6659,12 +6666,11 @@ def concatenate(arrays, axis=0):
         return data
     # OK, so we have to concatenate the masks
     dm = np.concatenate([getmaskarray(a) for a in arrays], axis)
+    dm = dm.reshape(d.shape)
```
Comment (eric-wieser) [inline numpy/ma/core.py]: No idea why this is needed, but it was here before too, and changing that is out of scope for this fix...

**Review by mhvk — APPROVED** [review]
Looks all OK, and good to see that common code factored out.

File: numpy/ma/core.py [hunk 2]
```diff
@@ -1925,7 +1934,7 @@ def masked_where(condition, a, copy=True):
         cls = MaskedArray
     result = a.view(cls)
     # Assign to *.mask so that structured masks are handled correctly.
-    result.mask = cond
+    result.mask = _shrink_mask(cond)
```
Comment (ahaldane) [inline numpy/ma/core.py]: So this function is the only real changed behavior, all the other changes are a factorization.

LGTM.

**Review by ahaldane — APPROVED** [review]

**Comment by ahaldane** [comment]
I'll go ahead with the merge, thanks Eric.

Not sure if backport is really needed.

**Comment by charris** [comment]
> Not sure if backport is really needed.

I really don't want to do any more backports at this point, instead I'd like to branch 1.14.x in a couple of weeks.
